lib/FrontendTool/ImportedModules.cpp

```diff
@@ -11,43 +11,99 @@
 //===----------------------------------------------------------------------===//
 
 #include "ImportedModules.h"
+#include "swift/AST/ASTContext.h"
 #include "swift/AST/Decl.h"
 #include "swift/AST/DiagnosticEngine.h"
 #include "swift/AST/DiagnosticsFrontend.h"
 #include "swift/AST/Module.h"
+#include "swift/Basic/STLExtras.h"
+#include "swift/ClangImporter/ClangImporter.h"
 #include "swift/Frontend/FrontendOptions.h"
+#include "clang/Basic/Module.h"
 #include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/StringRef.h"
 #include "llvm/Support/FileSystem.h"
 
-bool swift::emitImportedModules(DiagnosticEngine &diags, ModuleDecl *module,
+using llvm::StringRef;
+
+static StringRef getTopLevelName(const clang::Module *module) {
+  return module->getTopLevelModule()->Name;
+}
+
+static void findAllClangImports(const clang::Module *module,
+                                llvm::SetVector<StringRef> &modules) {
+  for (auto imported : module->Imports) {
+    modules.insert(getTopLevelName(imported));
+  }
+
+  for (auto sub : swift::makeIteratorRange(module->submodule_begin(),
+                                           module->submodule_end())) {
+    findAllClangImports(sub, modules);
+  }
+}
+
+bool swift::emitImportedModules(ASTContext &Context, ModuleDecl *mainModule,
                                 const FrontendOptions &opts) {
 
   auto path = opts.getSingleOutputFilename();
   std::error_code EC;
   llvm::raw_fd_ostream out(path, EC, llvm::sys::fs::F_None);
 
   if (out.has_error() || EC) {
-    diags.diagnose(SourceLoc(), diag::error_opening_output, path, EC.message());
+    Context.Diags.diagnose(SourceLoc(), diag::error_opening_output, path,
+                           EC.message());
     out.clear_error();
     return true;
   }
 
-  llvm::SmallVector<Decl *, 32> Decls;
-  module->getDisplayDecls(Decls);
+  llvm::SetVector<StringRef> Modules;
 
-  llvm::SetVector<Identifier> Modules;
-  for (auto *D : Decls) {
+  // Find the imports in the main Swift code.
+  llvm::SmallVector<Decl *, 32> Decls;
+  mainModule->getDisplayDecls(Decls);
```

Comment:
If you're thinking of the `Imports` field, it seems to only be filled by name binding, which I think this pass can't run (would require being able to resolve the dependencies).